[Merged by Bors] - chore(RingTheory): improve and generalize submodule localization API

alreadydone opened the pull request:
In Algebra/Module/LocalizedModule/Submodule, we introduce `Submodule.localized₀`, which is localization of a submodule considered as a submodule over the base ring rather than the localization. This allows us to talk about the localization of a submodule without choosing a specific ring localization, just like what `IsLocalizedModule` allows us to do for localization of a module.

As applications, `Rₚ` and every hypothesis depending on it are completely removed from the statement of `Submodule.le_of_localization_maximal`, etc. in RingTheory/LocalProperties/Submodule. We also reorder lemmas in this file to make the development more natural and golf the proofs (making use of `Module.eqIdeal` introduced in RingTheory/Ideal/Defs; also add some trivial lemmas.

In RingTheory/LocalProperties/Projective, we fix a proof due to removed explicit argument `Rₚ` and remove some unnecessary lines.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

erdOne commented on Mathlib/Algebra/Module/LocalizedModule/Submodule.lean:
```diff
@@ -28,26 +28,36 @@ Results about localizations of submodules and quotient modules are provided in t
+    rintro ⟨m, hm, s, hx⟩
+    exact ⟨r • m, smul_mem M' _ hm, s, by rw [IsLocalizedModule.mk'_smul, hx]⟩
+
+/-- Same as `Submodule.localized₀`, but viewed as an `S`-submodule. -/
```
Can you copy the docstring of `Submodule.localized₀` here?

erdOne commented on Mathlib/Algebra/Module/LocalizedModule/Submodule.lean:
```diff
@@ -28,26 +28,36 @@ Results about localizations of submodules and quotient modules are provided in t
 
+section Submodule
+
 /-- Let `S` be the localization of `R` at `p` and `N` be the localization of `M` at `p`.
```
There's no `S` at all in this declaration right?

erdOne reviewed (approved):
LGTM otherwise

Edit: other than the build error, that is

alreadydone commented on Mathlib/RingTheory/Ideal/Defs.lean:
```diff
@@ -68,6 +68,14 @@ theorem unit_mul_mem_iff_mem {x y : α} (hy : IsUnit y) : y * x ∈ I ↔ x ∈
 end Ideal
 
+/-- For two elements `m` and `m'` in an `R`-module `M`, the set of elements `r : R` with
+equal scalar product with `m` and `m'` is an ideal of `R`. -/
```
```suggestion
equal scalar product with `m` and `m'` is an ideal of `R`. If `M` is a group, this coincides with the
kernel of `LinearMap.toSpanSingleton R M (m - m')`. -/
```

Ruben-VandeVelde commented:
I feel like this would be much easier to review if it was split into more focused PRs. Probably each bullet point in your summary could be a separate PR.

alreadydone commented:
[This](https://github.com/leanprover-community/mathlib4/pull/19118/files/9411282f586957e4a54470e4d352fe9cbc4f4950..1ef9d7321d796b51b23e80d7208361c5c8a82975) is the diff after Andrew's approval, and I agree the changes in [Mathlib/RingTheory/Ideal/Defs.lean](https://github.com/leanprover-community/mathlib4/pull/19118/files/9411282f586957e4a54470e4d352fe9cbc4f4950..1ef9d7321d796b51b23e80d7208361c5c8a82975#diff-b07d9879de8e5aaf8c99da0c8f04b1eed23daf73af82696409093926ee1f3afe) could be another PR (depending on this one). In [RingTheory/LocalProperties/Submodule.lean](https://github.com/leanprover-community/mathlib4/pull/19118/files/9411282f586957e4a54470e4d352fe9cbc4f4950..1ef9d7321d796b51b23e80d7208361c5c8a82975#diff-25704369c7e20d4992fa2ef4a09fc372db1df67c0e7c504d4a99820bb1ec8f46) (which is the main file I changed and Andrew reviewed) I merely reordered the lemmas to partially restore the original order, extracted `Module.eqIdeal` from two proofs, and added trivial lemmas about `= ⊤` and equality of linear maps.

alreadydone commented:
@erdOne I hope this PR doesn't fall off your radar after I renamed the title :) If you think any part of it should be split, just let me know.

jcommelin commented:
If it isn't too hard to split according to the bullet points in OP, I would appreciate that.

alreadydone commented:
I discarded the part irrelevant to [#19080](https://github.com/leanprover-community/mathlib4/pull/19080#issuecomment-2480283549). This part has mostly been reviewed by @erdOne apart from a few later additions/refactors, and would be unnatural to break down further.

jcommelin commented:
Thanks :tada:

bors merge

alreadydone commented:
Forgot to mention: #19310 is the split-off part of this PR.
